dart/dynamics/Skeleton.cpp

```diff
@@ -8538,6 +8580,82 @@ Eigen::MatrixXs Skeleton::finiteDifferenceBodyLocalAccelerationsJacobian(
   return result;
 }
 
+//==============================================================================
+/// This computes the jacobian of the local gravity vectors for each body with
+/// respect to `wrt`
+Eigen::MatrixXs Skeleton::getBodyLocalGravityVectorsJacobian(
+    neural::WithRespectTo* wrt)
+{
+  int dim = wrt->dim(this);
+  Eigen::MatrixXs jac = Eigen::MatrixXs::Zero(getNumBodyNodes() * 3, dim);
+
+  if (wrt == neural::WithRespectTo::POSITION)
+  {
+    std::vector<BodyNode*>& bodyNodes = mSkelCache.mBodyNodes;
+
+    const Eigen::MatrixXi& parents = getJointParentMap();
+    for (int b = 0; b < bodyNodes.size(); b++)
+    {
+      int parentJointIndex
+          = bodyNodes[b]->getParentJoint()->getJointIndexInSkeleton();
+
+      // const math::Jacobian& H = getJacobianInPositionSpace(bodyNodes[b]);
+      Eigen::Isometry3s Tworld = bodyNodes[b]->getWorldTransform();
+      for (auto i = 0u; i < getNumDofs(); ++i)
+      {
+        const dynamics::DegreeOfFreedom* dof = getDof(i);
+        const dynamics::Joint* joint = dof->getJoint();
+        if (joint->getJointIndexInSkeleton() == parentJointIndex
+            || parents(joint->getJointIndexInSkeleton(), parentJointIndex) == 1)
+        {
+          Eigen::Vector3s worldRotationScrew
+              = joint->getWorldAxisScrewForPosition(dof->getIndexInJoint())
+                    .head<3>();
+          Eigen::Vector3s localRotationScrew
+              = Tworld.linear().transpose() * worldRotationScrew;
+          // Derivative of gravity acceleration
+          Eigen::Vector3s localGravity
+              = Tworld.linear().transpose() * getGravity();
+          jac.block<3, 1>(b * 3, i)
+              = -1 * localRotationScrew.cross(localGravity);
+        }
+      }
+    }
+  }
+  return jac;
+}
+
+//==============================================================================
+/// This brute forces our local gravity vectors jacobian
+Eigen::MatrixXs Skeleton::finiteDifferenceBodyLocalGravityVectorsJacobian(
+    neural::WithRespectTo* wrt)
+{
+  int dim = wrt->dim(this);
+  Eigen::MatrixXs result(getNumBodyNodes() * 3, dim);
+  Eigen::VectorXs original = wrt->get(this);
+
+  s_t eps = 1e-3;
+  bool useRidders = true;
+  math::finiteDifference(
+      [&](/* in*/ s_t eps,
+          /* in*/ int dof,
+          /*out*/ Eigen::VectorXs& perturbed) {
+        Eigen::VectorXs tweaked = original;
+        tweaked(dof) += eps;
+        wrt->set(this, tweaked);

```

Comment:
I'm not totally sure how the `neural::WithRespectTo` stuff works. It's popping up more and more, so I should probably get familiar with it.

Reply:
I think this is worth a larger in-person conversation, but my brief summary is that `neural::WithRespectTo` is passed as an argument to methods that might want to return a gradient or Jacobian with respect to a few different things, where the structure of the calculation is basically the same, except for some small inner details.

It also has functionality to allow finite differencing against a `neural::WithRespectTo`, to make testing easier. For example, `neural::WithRespectTo::POSITION->set(skel, pos)` will just set the position of `skel` to `pos`.